### PR TITLE
bpo-35498: Added slice support to PathLib parents attribute.

### PR DESCRIPTION
**Pull request by thejcannon**
https://bugs.python.org/issue35498

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Our records indicate we have not received your CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for your contribution, we look forward to reviewing it!

**Comment by terryjreedy** [comment]
On the issue, Joshua wrote "CLA signage is pending approval at the company I work for, " .

File: Lib/pathlib.py
Comment (eamanu) [inline Lib/pathlib.py]: What about use `self[_] for _ in range ...`

**Review by eamanu — APPROVED** [review]
LGTM.

File: Lib/pathlib.py
Comment (thejcannon) [reply Lib/pathlib.py]: I believe from experiences in the Python world that using underscore as a variable name implies you won't be using it at all.

**Comment by thejcannon** [comment]
CLA has been signed 😃

**Comment by JulienPalard** [comment]
See the issue: I'm trying to understand in which case it could be usefull to slice parents, please comment on the issue.

File: Lib/test/test_pathlib.py
Comment (pganssle) [inline Lib/test/test_pathlib.py]: How about a test for iterating over the list backwards? `par[::-1]`?

File: Lib/pathlib.py
Comment (pganssle) [inline Lib/pathlib.py]: Since `PathLib.parents` returns an immutable sequence, slices of it should also be immutable, IMO, so this should return a tuple:

```suggestion
            return tuple(self[i] for i in range(*idx.indices(len(self))))
```

File: Misc/NEWS.d/next/Library/2018-12-14-13-29-17.bpo-35498.LEJHl7.rst
Comment (pganssle) [inline Misc/NEWS.d/next/Library/2018-12-14-13-29-17.bpo-35498.LEJHl7.rst]: Better to refer to `:meth:~pathlib.Path.parents`, I believe, since that is the documented function.

**Review by pganssle — CHANGES_REQUESTED** [review]
I think that `pathlib.Path.parents` should support slices and negative indexing.

Assuming @pitrou doesn't object, I am thinking that we'll merge #21799 (since that's a much simpler PR), then this should be rebased onto that PR, and the `.. versionchanged::` and What's New entries should be updated.

File: Lib/pathlib.py
Comment (pganssle) [inline Lib/pathlib.py:632]: By the way, another option to consider (and this is a bit unusual so I will defer to others if there's an objection) would be to use `@functools.singledispatchmethod` for handling the slices in what may be a neater way:

```python
@functools.singledispatchmethod
def __getitem__(self, idx):
    if idx < 0 or idx >= len(self):
        raise Indexerror(idx)
    return self._pathcls._from_parsed_parts(self._drv, self._root,
                                            self._parts[:-idx - 1])

@__getitem__.register(slice)
def _(self, idx):
    return tuple(self[i] for i in range(*idx.indices(len(self))))
```

I tend to like single dispatch rather than multiple returns, but it might be weird in a dunder method. Plus, I can find no other uses of `@functools.singledispatchmethod` anywhere else in the standard library, so maybe you won't want to rock the boat in that regard 😛.

Comment (thejcannon) [reply Lib/pathlib.py:632]: I like that better for sure, but I'll wait to see if others have opinions.

Comment (pitrou) [reply Lib/pathlib.py:632]: FTR, using `singledispatch` for this looks awkward to me.
(but perhaps pattern matching would help... ;-))

Comment (thejcannon) [reply Lib/pathlib.py:632]: Also, I think there's some precedence:
https://github.com/python/cpython/blob/8d59eb1b66c51b2b918da9881c57d07d08df43b7/Lib/tracemalloc.py#L204
https://github.com/python/cpython/blob/9f55551f3df238e58315e724e50cb0d574d75b94/Lib/sre_parse.py#L165
https://github.com/python/cpython/blob/85339f5c220a5e79c47c3a33c93f1dca5c59c52e/Lib/calendar.py#L60

Comment (pganssle) [reply Lib/pathlib.py:632]: `functools.singledispatchmethod` was introduced in Python 3.8 I think. Those other implementations are much older and wouldn't have had this option (they may not have even had the option of `functools.singledispatch`).

If Antoine doesn't like it I'm fine with leaving it with a simple `isinstance` check (though hopefully this doesn't get used as precedent for "we never do that in CPython", since I think everyone involved is pretty ambivalent or apathetic about it 😛).

**Comment by thejcannon** [comment]
@pganssle I went ahead and updated the What's New and pathlib documentation (I'll resolve any merge issues after #21799 is in, but I wanted to resolve the issues while this is fresh),

I'm excited to add "Contributed to Python" to my resume, even if my contribution is small 😄

**Comment by thejcannon** [comment]
I have made the requested changes; please review again

**Review by pganssle — APPROVED** [review]

**Comment by pganssle** [comment]
OK, since this one is ready first, I'll merge it and the other one can be rebased on top of master.

Thanks for your work on this @thejcannon, and thanks for the quick turnaround on post-review changes after your PR has been sitting idle for almost 2 years!

**Comment by thejcannon** [comment]
@pganssle Happy to help <3

**Comment by hroncok** [comment]
Thanks for this!

**Pull request merged**
